source-mysql-batch/main.go

```diff
@@ -80,24 +80,35 @@ func connectMySQL(ctx context.Context, cfg *Config) (*client.Conn, error) {
 	}).Info("connecting to database")
 
 	var conn *client.Conn
+
+	const mysqlErrorCodeSecureTransportRequired = 3159 // From https://dev.mysql.com/doc/mysql-errors/8.4/en/server-error-reference.html
+	var mysqlErr *mysql.MyError
 	var withTLS = func(c *client.Conn) error {
 		c.SetTLSConfig(&tls.Config{InsecureSkipVerify: true})
 		return nil
 	}
+	// The following if-else chain looks somewhat complicated but it's really very simple.
+	// * We'd prefer to use TLS, so we first try to connect with TLS, and then if that fails
+	//   we try again without.
+	// * If either error is an incorrect username/password then we just report that.
+	// * Otherwise we report both errors because it's better to be clear what failed and how.
+	// * Except if the non-TLS connection specifically failed because TLS is required then
+	//   we don't need to mention that and just return the with-TLS error.
 	if connWithTLS, errWithTLS := client.Connect(cfg.Address, cfg.User, cfg.Password, cfg.Advanced.DBName, withTLS); errWithTLS == nil {
 		log.WithField("addr", cfg.Address).Info("connected with TLS")
 		conn = connWithTLS
+	} else if errors.As(errWithTLS, &mysqlErr) && mysqlErr.Code == mysql.ER_ACCESS_DENIED_ERROR {
+		return nil, cerrors.NewUserError(mysqlErr, "incorrect username or password")
 	} else if connWithoutTLS, errWithoutTLS := client.Connect(cfg.Address, cfg.User, cfg.Password, cfg.Advanced.DBName); errWithoutTLS == nil {
 		log.WithField("addr", cfg.Address).Info("connected without TLS")
 		conn = connWithoutTLS
+	} else if errors.As(errWithoutTLS, &mysqlErr) && mysqlErr.Code == mysql.ER_ACCESS_DENIED_ERROR {
+		log.WithField("withTLS", errWithTLS).Warn("connecting with TLS failed for an unrelated reason")
```

Comment:
nit: I would reword this a little bit to avoid ambiguity for users about the "unrelated" error, although my suggestion is a bit of a mouthful, feel free to use whatever you think suits
```suggestion
		log.WithField("withTLS", errWithTLS).Warn("connecting with TLS failed for a non-TLS-related reason")
```

Reply:
That's a good point, though it should probably be addressed in the other PR -- this one just includes that diff because I wanted to get the merge conflict between the two resolved ahead of time.